New duplicate algorithm to check for similar entries

george-gca opened the pull request:
I added the option to check for duplicate entries based on the similarity of the title and abstract. Sometimes we can have a duplicate entry that is a fixed version of another entry, with a corrected typo or added comma, for example.

I decided to go with [difflib.SequenceMatcher](https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher) for this similarity, since it is a built-in solution. Added the options to use only the title for this check, set the similarity score, discard stopwords for a more strict check considering only the useful words, and also added a pretty diff print support thanks to the [rich](https://github.com/Textualize/rich) library:

![dedup_similar](https://github.com/user-attachments/assets/2a5022fe-9e48-423f-ad01-7e47148051eb)

george-gca committed:
Added more dependencies to project

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Added new duplicate finding algorithm

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Added more params to deduplicate similar function

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Added missing params for deduplicate similar entries

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Updated README

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Added similar dedup info to Tutorials

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Added missing example image

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Fixed ruff warnings

Signed-off-by: George Araújo <george.gcac@gmail.com>

PeterLombaers commented:
Thanks for this contribution, it look really nice! It makes a lot of sense to me to want to deduplicate using some fuzzy matching. The code looks clean! I was testing out the features and it seems to be working well. Just a few comments, which I'll put in the comments below.

PeterLombaers commented:
When I was testing with `--verbose`, I got confused when I did not see anything pretty printing. I think this happened because my duplicated were already dropped by the original dedup method. When I added rows that were dropped only by your method, I indeed so the nice diff. If I'm using `--verbose` I expect to get feedback no matter which method does the deduplicating.

PeterLombaers commented on asreviewcontrib/datatools/entrypoint.py:
```diff
@@ -59,6 +60,43 @@ def execute(self, argv):
+                    action='store_true',
+                    help="Drop similar records.",
+                )
+                dedup_parser.add_argument(
```
Maybe call this `threshold`?

george-gca replied:
The comments came in the wrong place I think. You mean the `similarity` right?

PeterLombaers commented on asreviewcontrib/datatools/entrypoint.py:
```diff
@@ -59,6 +60,43 @@ def execute(self, argv):
+                    action='store_true',
+                    help="Use only title for deduplication.",
+                )
+                dedup_parser.add_argument(
```
Maybe call this `skip_stopwords`, so that it's more similar to the command above?

george-gca replied:
I was thinking of not removing the stopwords by default. Do you think it's best to leave it on by default?

george-gca replied:
I was thinking of not removing the stopwords by default. Do you think it's best to leave it on by default?

PeterLombaers replied:
I agree, removing stopwords should not be on by default. My point was more that I thought the names are a bit confusing. There is `drop`, `skip` and `discard`, which all mean very similar things.

george-gca replied:
What about:

- `--drop_similar` -> `--similar`
- `--skip_abstract` -> `--title_only`
- `--discard_stopwords` -> `--stopwords`
- `--strict_similarity` -> `--strict`

PeterLombaers replied:
Yes, I like it!

george-gca replied:
Just applied these changes to the latest commit.

PeterLombaers commented on asreviewcontrib/datatools/entrypoint.py:
```diff
@@ -59,6 +60,43 @@ def execute(self, argv):
+                    action='store_true',
+                    help="Discard stopwords for deduplication.",
+                )
+                dedup_parser.add_argument(
```
I also find this name and help description a bit confusing. It's about the similarity technique you use, right?

george-gca replied:
Yes, this is to enable the use of the `ratio()` function along with the `real_quick_ratio()` and `quick_ratio()`. But I don't think the user has to understand these details, only that it will use a more stricter check.

PeterLombaers commented:
But thanks again, it looks very nice! We'll also need to have a chat about how this relates to asreview2.0 @J535D165 !

george-gca commented:
What do you think should be done when verbose is true for the current algorithm? I mean, what would be the expected output?

george-gca committed:
Renamed similarity as threshold

Signed-off-by: George Araújo <george.gcac@gmail.com>

Rensvandeschoot commented:
Great contribution!! I have recently used fuzzy matching for our [FORAS project](https://www.crd.york.ac.uk/prospero/display_record.php?RecordID=494027), where we obtained over 10K records to screen. I checked for duplicates within the dataset and between the titles obtained via the database search and the most likely title match in OpenAlex. I saved the matching score and went through the data, checking these scores from low to high, and I found many fuzzy duplicates of the following type:

•	Titles containing extra or different punctuation.
•	One title has a spelling mistake corrected in the other.
•	Presence of HTML code in one title (e.g., <i>PTSD</i>), but not in the other.
•	An abstract number at the beginning of one title (e.g., “T262.”), missing from the other.
•	A subtitle in one record versus a single-title format in the other.

All such cases are precise duplicates and can be corrected without losing any information. 

But I also found cases with different versions of the same work:
- re-prints of the same paper in a different journal,
- pre-print + journal version, 
- conference abstract + journal version, 
- book chapter + journal version, 
- dissertation + journal version, 
- version 1 + version 2 of the same paper

You might want to keep both records in these cases, but the labeling decisions will be the same.

So, my question is whether it is possible to store the matching score so that a user can manually check the records with lower matching scores? 

and, hopefully, my comment helps with starting the discussion on what to do with fuzzy-duplicates in ASReview v2.0 :-)

george-gca commented:
Currently you can choose to print the duplicates on the terminal instead of already removing the duplicates by avoiding the `-o` option. It will print the line of the entries and a pretty diff between them, but it will print this in order of finding, not based on the score. This could be added as an option, for instance, or added as a column to the dataset like you said. What option would be best? Also, adding this column to the dataset wouldn't somehow affect the usage of the dataset with ASReview, or does it simply ignore the extra columns?

This should be enough to match most of the cases you pointed here just by playing with the `threshold` param, since I do [some cleaning](https://github.com/george-gca/asreview-datatools/blob/a27bad91ff1ffc6d9bc253c6e67f6f9f0b43e9c3/asreviewcontrib/datatools/dedup.py#L59-L64) before doing the actual match. I've also added an option to [remove stopwords](https://github.com/george-gca/asreview-datatools/blob/a27bad91ff1ffc6d9bc253c6e67f6f9f0b43e9c3/asreviewcontrib/datatools/dedup.py#L77) before checking, to create even more similar titles. The cases that could fall short would be:

- presence/abscence of HTML or latex code, if it is too long
- added subtitle to the title

The added subtitle might be a pitfall, since some papers build on top of others, including the title, being completely different papers with similar titles. And sometimes they like to follow a trend of titles, like the case for [x is all you need](https://github.com/vinayprabhu/X-is-all-you-need).

It would be actually a great test for this code the project you mentioned.

PeterLombaers commented:
> What do you think should be done when verbose is true for the current algorithm? I mean, what would be the expected output? Because the pretty print probably will print everything dimmed in most cases.

True, I'm not sure what I would want to see in verbose mode for the other case. I would leave it as it is for now. We might to think in the future about wether we want to maintain this feature and how it would look then, also for the other deduplication. I do think it's very nice to have such verbose output when deduplicating, so that you can clearly see which ones get marked as duplicate and why.

george-gca committed:
Printing when not using similar

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca commented:
I agree with you. I had a quick look at the [ASReviewData.drop_duplicates](https://asreview.readthedocs.io/en/stable/_modules/asreview/data/base.html#ASReviewData.drop_duplicates) code, and it resets index by default, meaning that comparing dataframes before and after the deduplication would fail.

I changed the code a little bit to allow verbose when not using similar, also moved part of the code to `dedup.py` since it was starting to be too much code in the entrypoint for that.

PeterLombaers commented on asreviewcontrib/datatools/dedup.py:
```diff
@@ -102,4 +103,59 @@ def drop_duplicates_by_similarity(
+            if len(duplicate_entries) > 0:
+                print("Duplicate entries:")
+                for i, row in duplicate_entries.iterrows():
+                    print(f"\tLine {i} - {row['title']}")
```
Maybe print the PID as well here? It is also deduplicating based on pid after all.

george-gca replied:
Sure, I will add it to it.

PeterLombaers commented on asreviewcontrib/datatools/dedup.py:
```diff
@@ -76,7 +77,7 @@ def drop_duplicates_by_similarity(
         s = s.str.replace(stopwords_regex, '', regex=True)
 
-    duplicated = (s.duplicated()) & (s.notnull())
+    duplicated = [False] * len(s)
```
I prefer `[False for _ in range(len(s))]`. In this case it does not really matter, but if you put a dictionary instead of `False` inside the list, then your list will actually contain references to _the same_ dictionary instead of copies of dictionaries.

george-gca replied:
Yes, but in this case the content is just `False`, and in the line 101 this value will be updated to `True`, there will be never a dictionary here. And in this case, the current solution is faster. See:

```python
In [1]: %timeit [False] * 10_000_000
29.6 ms ± 2.81 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [2]: %timeit [False for _ in range(10_000_000)]
187 ms ± 2.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

PeterLombaers commented on asreviewcontrib/datatools/dedup.py:
```diff
@@ -102,4 +103,59 @@ def drop_duplicates_by_similarity(
+            "because there is no such data."
+        )
+
+    if not args.similar:
```
Now you can't use both kinds of deduplication at the same time. Did you have a reason for that? I think it would be easier to use if you can just do `asreview data dedup file.csv --similar` and you get a fully deduplicated set, instead of having to run deduplication twice, once with the `--similar` flag and once without.

george-gca replied:
If you run with the similar flag it will also catch the cases where it is an exact duplicate, since the similarity score will be higher than of a just similar case.

PeterLombaers replied:
But you will not catch duplicates based on PID, right?

george-gca replied:
When checking for similar papers it won't, but that's kind of the point, isn't it?

PeterLombaers replied:
If I want to remove duplicates based on PID and on similar title/abstract, I have to call `asreview data dedup file.csv` and then call `asreview data dedup file.csv --similar`. I just thought it'd might be nice to call `asreview data dedup file.csv --similar` and get both kinds of deduplication in one go. But if you like it better if they are separate, I'm also fine with it. I don't have a strong opinion.

george-gca replied:
It is not that I am against it, it is more like, if the pid is different it is high likely that the title won't be the same. But maybe this could find cases where the pid is the same but the titles completely different, and warn the user about it. I will include it in the solution as well.

PeterLombaers reviewed (commented):
I like the changes! I just have a few comments, but other than that I'm all for merging :)

george-gca committed:
Printing PID of duplicate entries

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca commented:
One question, is it really necessary to enforce the maximum line length to be 88? I mean, this line length is too small even for someone working on a laptop screen.

![image](https://github.com/user-attachments/assets/6948b566-e72f-4960-bc75-03188a074b9e)

For instance, line 124 has 88 columns, and it doesn't even reach half of my laptop screen (it has 1920x1080 resolution). I feel like some print statements in this code would be more readable if this line length increased by a little, and it wouldn't hurt the coding experience.

PeterLombaers commented:
I kinda like it, I can put two windows next to each other on my laptop.

PeterLombaers commented:
![image](https://github.com/user-attachments/assets/1f38dde4-1146-4fc3-92e0-ea1edd828fae)
This is my usual setup, which actually fits on my laptop, so that's why I like it. But feel free to file an issue on the asreview repo and see if you can convince @J535D165 😄.

george-gca commented:
Even in your setup the lines trespass the 88 columns threshold for half monitor and you need to scroll right to see the rest of it, so a line wrap is more fitting I think. But that is just personal opinion.

george-gca commented:
@PeterLombaers do you have any dataset with DOIs that I can use to test the new solution? The change is rather annoying if I am to consider pids for the similar case, like this loop has to change:

```python
for j, t in s.iloc[i+1:][abs(s.str.len() - len(text)) < 5].items():
```

george-gca committed:
Also supporting PID when searching for similar

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca commented:
I believe now it is finished.

PeterLombaers commented:
Sorry, I feel like there was some miscommunication! In your last commit you implement deduplicating based on PID, but `ASReviewData.drop_duplicates` already does this. So having a second implementation will only confuse matters. 

How about the following plan to keep things simple:
- Revert the last commit
- If `--similar`, we use your method
- Else we `ASReviewData.drop_duplicates`, so simply how it is in the current version of datatools
- If you want to use both, you have to call `asreview data dedup` twice, once with `--similar` and once without
- Any extra arguments like `--verbose` and `--title_only`, etc. will only apply in case of the `--similar` flag
- in the help text of the arguments it says whether the argument only applies for `--similar`
- We add a test to check that the different options give the correct deduplication

This means that the behaviour with `--similar` and without is not quite the same. For example, if you don't use `--similar`, you always use both title and abstract for deduplicating, but if you use `--similar`, you can select `--title_only`. But I'm fine with that difference for now, lets get you PR into shape to merge, then we can always look later at streamlining the datatools API as a whole.

If you do the first five points on that list, I can help by writing tests, adding the help texts and linting the code.

Do you agree with this plan?

PeterLombaers commented:
BTW, When checking out your pull request, I made the following csv file to test the different settings:
```
abstract, title, doi, publication_year
abstract, <h2>this is a title of a paper<h2>, doi1, 2005
abstract, this is a title of a paper, doi1, 2001
abstract, this is a title of a paper&, doi3,
abstract, this is the title of a paper, doi4, 2003
abstract, this is a title%^ of a paper, doi5, 2004
abstract, this is a title of a pãper, doi6, 2000
```
Might be useful for running sanity checks that everything is working as expected.

george-gca commented:
> If you do the first five points on that list, I can help by writing tests, adding the help texts and linting the code.
> 
> Do you agree with this plan?

Sorry, I disagree. I agree with what you previously said (or at least that's what I understood), that it is best to have a streamlined functionality that looks like one single tool, not like two separate tools were glued together. Also one should not need to run once without `--similar` and once with `--similar`, since an exact copy is also a similar considering the string difference algorithm.

I have another proposition: make all output look like the version with `--similar` flag. The difference is that without the similar flag we will look for exact matches kinda like it is currently done, only making a few changes so I can find which one is the original and which is the duplicate, and keep it how it is for the `--similar` case. I won't suggest always using the same algorithm because without the `--similar` is way faster.

What I am thinking now for the case without `--similar` is:
- remove duplicates, compare sizes before and after removal
- size changed (meaning some duplicates were removed), check again for duplicates in the original data (now with [keep=False](https://pandas.pydata.org/docs/reference/api/pandas.Series.duplicated.html#pandas-series-duplicated) so we flag all data with copies, not only the duplicates) and create a df with only the copies
- use the same algorithm used in the `--similar` case, but only for this subset of the data that have/are copies. This way we can find from which exactly row that duplicate is a duplicate of, and running in a potentially smaller subset will be faster
- use the same print as for the similar case

What do you think? Also, help for the tests and linting would be great.

PeterLombaers commented:
The reason I suggest keeping it split, is that in the near future (a few months), ASReview version 2 will be released. The `ASReviewData` object will change and how we deal with duplicates inside ASReview might also change. The current deduplicate without `--similar` makes use of `ASReviewData.drop_duplicates`, so if we change that now we will have to do the same work again for ASReview v2.

If we keep the two methods (with and without `--similar`) separate for now, we can unify them once v2 lands. The `--similar` method can be completely inside `asreview-datatools`, and I don't mind making a new release for the updated dedup method for datatools. But I don't think we want any more releases for ASReview v1 that are not bugfixes.

george-gca commented:
That makes sense. So what do you think I should do? Wait for V2?

PeterLombaers commented:
I leave the choice up to you. We can add it to v1 with the plan I wrote above, or we leave this pull request open for now and then integrate your deduplication algorithm in the v2 deduplication. Then we'll be careful that you get the correct recognition as a contributer when we implement the deduplication in v2.

george-gca commented:
I can keep them separate for now and get the PR accepted so some users (including me) might benefit from it. Then, when V2 launches I can get back to it and port the solution. What do you think?

george-gca commented:
I won't revert it exactly cause I don't want to lose the code (it will be useful for the port), but I will make a new commit undoing stuff.

george-gca committed:
Changed dedup without --similar to use old solution

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca committed:
Created tests for dedup

Signed-off-by: George Araújo <george.gcac@gmail.com>

george-gca commented:
Just created a few tests for dedup. Give it a try @PeterLombaers.

Also, kind of not related, but is there any update regarding [1507](https://github.com/asreview/asreview/issues/1507)? I mean, will this be changed for v2?

PeterLombaers commented:
Thanks for the changes, the old and new flow are clearly separated. And thanks for adding the tests. When I tried running them they failed though. The problem is that the `deduplicate_data` function takes the complete `args` objects as input. That makes it quite difficult to directly call the function, since you have to recreate all the default values that argparse  normally creates for you.

Instead of letting you fix the tests, I'm going to go ahead and merge the pull request. I'll fix the tests and do the linting in a separate pull request after. That way I can speed up things a bit, because as you probably noticed I don't have a lot of time to review your changes during the week.

PeterLombaers commented:
Thanks again for the pull request. It's nice to see people from all over contributing to ASReview!
